EXTENSIONS.md

```diff
@@ -0,0 +1,113 @@
+# Linkerd Extensions
+
+Linkerd has an extension model which allows 3rd parties to add functionality.
+Each extension consists of a binary CLI named `linkerd-name` (where `name` is
+the name of the extension) as well as a set of Kubernetes resources.  To invoke
+an extension, users can run `linkerd name` which will search the current PATH
+for an executable named `linkerd-name` and execute it.
+
+## Installing Extensions
+
+To install an extension, first download the extension executable and put it
+in your PATH. The extension can then be installed into your Kubernetes cluster
+by running `linkerd <extension name> install | kubectl apply -f -`. Similarly,
+it can be uninstalled by running
+`linkerd <extension name> uninstall | kubectl delete -f -`.
+
+A full list of installed extensions can be printed by running `linkerd check`.
+
+## Developing Extensions
+
+The extension must be an executable file named `linkerd-name` where `name` is
+the name of the extension. The name must not be any of the built-in Linkerd
+commands (e.g. `check`) or extensions (e.g. `viz`).
+
+The extension must accept the following flags and respect them any time that
+it communicates with the Kubernetes API.  All of these flags must be accepted
+but may be ignored if they are not applicable.
+
+* `--api-addr`: Override kubeconfig and communicate directly with the control plane at host:port (mostly for testing)
+* `--context`: Name of the kubeconfig context to use
+* `--help`/`-h`: Print help message
+* `--kubeconfig`: Path to the kubeconfig file to use for CLI requests
+* `--linkerd-namespace`/`-l`: Namespace in which Linkerd is installed [$LINKERD_NAMESPACE]
+* `--verbose`: Turn on debug logging 
+
+The extension must implement these commands:
+
+### `linkerd-name install`
+
+This command must print the Kubernetes manifests for the extension as yaml
+which is suitable to be passed to `kubectl apply -f`. These manifests must
+include a Namespace resource with the label `linkerd.io/extension=name` where
+`name` is the name of the extension. This allows Linkerd to detect installed
+extensions.
```

Comment:
is the intent for linkerd to iterate over all namespaces, searching for the `linkerd.io/extension=name`? how does this set of extensions relate to the set of binaries detected via PATH?

Reply:
We use a label selector rather than iterating over all namespaces, but yes, that's the idea.  Installed extensions are discovered by listing all namespaces with the `linkerd.io/extension` label.  For each extension discovered in this way, we look for an executable on the path named `linkerd-<extension name>`.